### ISSUE TITLE
Feature for treating non-existent, relative files as external files

### ISSUE THREAD

**Issue by joeldenning** [rollup/rollup#532]
https://github.com/rollup/rollup/issues/410 and https://github.com/rollup/rollup/issues/418 (as explained in [the wiki](https://github.com/rollup/rollup/wiki/JavaScript-API#external)) make it possible to treat relative files as external by putting `path.resolve('./relative-dep.js')` into the `externals` array of your rollup configuration. However, as far as I can tell, that approach for skipping relative files will only work if the file exists (otherwise `path.resolve` will fail inside of the rollup configuration).

The use case for skipping a relative import without there being a file of the same name is for bundling projects that will later be `System.import`ed (using SystemJS). The reason why the import does not match the relative filename exactly is because the project specifies which SystemJS plugin to use when loading the relative import. For example, my file tree may look like this:
```
src
-->main.js
-->main.css
```

where main.js is
```
import './main.css!cssmodules';
...
```

Is there a way I am unware of that allows for you to skip `./main.css!cssmodules` when the filename is really just `./main.css`??

**Comment by Conduitry** [comment]
I've only been using Rollup for a few days, so take this all with a grain of salt.

`path.resolve()` does in fact work with non-existing files.

    > path.resolve('./main.css!cssmodules')
    '/home/whatever/main.css!cssmodules'

So I would guess you could just specify the ! suffix in what you pass to `pass.resolve()` for your `externals`.

If for some reason this doesn't work, or if you don't want to have to specify (or simply don't know) what all the different suffixes are going to be, you could write a plugin with a custom `resolveId` function. See https://github.com/rollup/rollup/wiki/Plugins under "Creating plugins". Something I've found interesting and helpful is to write a testing plugin with one of those methods given on that wiki page that simply `console.log()`s its arguments, and returns undefined. This (generally) won't affect Rollup's behavior, but it does let you see what path names etc. Rollup is looking at under the hood.

**Comment by joeldenning** [comment]
@Conduitry you are right that `path.resolve('main.css!css')` actually does not throw an error. But rollup still throws an error even when I declare `main.css!css` as an external. For example,

```js
// rollup.config.js
import path from 'path';

export default {
    entry: 'main.js',
    dest: 'bundle.js',
    external: [path.resolve('./main.css!css')],
}
```

```js
// main.js
import './main.css!css';
```

```js
// main.css
.mycss_class {
}
```

throws the following error (with rollup@0.25.4):
```
Error: Could not resolve ./main.css!css from /Users/joel/playground/test/main.js
    at /Users/joel/playground/test/node_modules/rollup/www/ROLLUP/rollup/src/Bundle.js:186:42
    at tryCatch (/Users/joel/playground/test/node_modules/rollup/www/ROLLUP/rollup/node_modules/es6-promise/lib/es6-promise/-internal.js:185:12)
    at invokeCallback (/Users/joel/playground/test/node_modules/rollup/www/ROLLUP/rollup/node_modules/es6-promise/lib/es6-promise/-internal.js:197:13)
    at publish (/Users/joel/playground/test/node_modules/rollup/www/ROLLUP/rollup/node_modules/es6-promise/lib/es6-promise/-internal.js:168:7)
    at flush (/Users/joel/playground/test/node_modules/rollup/www/ROLLUP/rollup/node_modules/es6-promise/lib/es6-promise/asap.js:87:5)
    at doNTCallback0 (node.js:407:9)
    at process._tickCallback (node.js:336:13)
    at Function.Module.runMain (module.js:477:11)
    at startup (node.js:117:18)
    at node.js:951:3
```

However, if I rename the `main.css` file to be `main.css!css`, then no error is thrown. To me, this seems like something that should be taken care of inside of Rollup itself (as opposed to a rollup plugin), since Rollup already handles `externals`, without plugins. Also, as far as I can tell, #410, #418, and [the wiki](https://github.com/rollup/rollup/wiki/JavaScript-API#external) did not intentionally create this behavior where the relative file must exist, which might indicate that this was just an oversight?

**Comment by joeldenning** [comment]
Ping.

If this is a feature that everyone's okay with, I'd be happy to work on a pull request for it.

**Comment by Rich-Harris** [comment]
Released as 0.25.7. Thanks!

**Comment by joeldenning** [comment]
@Rich-Harris Thanks, look forward to updating all our rollup projects to the latest version!. I have also opened up rollup/rollup-plugin-node-resolve#32 which addresses this issue for those using rollup-plugin-node-resolve as well.
